crates/core/tedge_agent/src/restart_manager/actor.rs

```diff
@@ -0,0 +1,171 @@
+use crate::restart_manager::config::RestartManagerConfig;
+use crate::restart_manager::error::RestartManagerError;
+use crate::restart_manager::restart_operation_handler::restart_operation::create_tmp_restart_file;
+use crate::restart_manager::restart_operation_handler::restart_operation::has_rebooted;
+use crate::state_repository::state::AgentStateRepository;
+use crate::state_repository::state::RestartOperationStatus;
+use crate::state_repository::state::State;
+use crate::state_repository::state::StateRepository;
+use crate::state_repository::state::StateStatus;
+use async_trait::async_trait;
+use tedge_actors::Actor;
+use tedge_actors::MessageReceiver;
+use tedge_actors::RuntimeError;
+use tedge_actors::Sender;
+use tedge_actors::SimpleMessageBox;
+use tedge_api::OperationStatus;
+use tedge_api::RestartOperationRequest;
+use tedge_api::RestartOperationResponse;
+use tedge_config::system_services::SystemConfig;
+use tokio::process::Command;
+use tracing::error;
+use tracing::info;
+
+const SYNC: &str = "sync";
+
+#[cfg(not(test))]
+const SUDO: &str = "sudo";
+#[cfg(test)]
+const SUDO: &str = "echo";
+
+pub struct RestartManagerActor {
+    config: RestartManagerConfig,
+    state_repository: AgentStateRepository,
+    message_box: SimpleMessageBox<RestartOperationRequest, RestartOperationResponse>,
+}
+
+#[async_trait]
+impl Actor for RestartManagerActor {
+    fn name(&self) -> &str {
+        "RestartManagerActor"
+    }
+
+    async fn run(&mut self) -> Result<(), RuntimeError> {
+        self.process_pending_restart_operation().await?;
+
+        while let Some(request) = self.message_box.recv().await {
+            if let Err(err) = self.handle_restart_operation(&request).await {
+                error!("{}", err);
+                self.handle_error(&request).await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl RestartManagerActor {
+    pub fn new(
+        config: RestartManagerConfig,
+        message_box: SimpleMessageBox<RestartOperationRequest, RestartOperationResponse>,
+    ) -> Self {
+        let state_repository = AgentStateRepository::new_with_file_name(
+            config.config_dir.clone(),
+            "restart-current-operation",
+        );
+        Self {
+            config,
+            state_repository,
+            message_box,
+        }
+    }
+
+    async fn process_pending_restart_operation(&mut self) -> Result<(), RestartManagerError> {
+        let state: Result<State, _> = self.state_repository.load().await;
+
+        let mut status = OperationStatus::Failed;
```

Comment:
I would avoid a mutable status and would prefer to explicitly return a status for each case.